compliance/index/v4/node.go

```diff
@@ -0,0 +1,182 @@
+package v4
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/quay/claircore"
+	ccindexer "github.com/quay/claircore/indexer"
+	"github.com/quay/claircore/indexer/controller"
+	"github.com/quay/claircore/rhel"
+	rpm2 "github.com/quay/claircore/rpm"
+	"github.com/quay/zlog"
+	"github.com/stackrox/rox/compliance/collection/compliance"
+	"github.com/stackrox/rox/compliance/collection/intervals"
+	v4 "github.com/stackrox/rox/generated/internalapi/scanner/v4"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/scannerv4/mappers"
+)
+
+// The layer carries a hardcoded digest, as it is exclusively used for passing
+// ClairCore checks, not for Scanners matching
+var layerDigest = fmt.Sprintf("sha256:%s", strings.Repeat("a", 64))
+
+type localNodeIndexer struct {
+}
+
+// NewNodeIndexer creates a new node indexer
+func NewNodeIndexer() compliance.NodeIndexer {
+	return &localNodeIndexer{}
+}
+
+// GetIntervals
+func (l *localNodeIndexer) GetIntervals() *intervals.NodeScanIntervals {
+	i := intervals.NewNodeScanIntervalFromEnv()
+	return &i
+}
+
+// IndexNode indexes a live fs.FS at the container mountpoint given in the basePath.
+func (l *localNodeIndexer) IndexNode(ctx context.Context) (*v4.IndexReport, error) {
+	report := &claircore.IndexReport{
+		Packages:      map[string]*claircore.Package{},
+		Environments:  map[string][]*claircore.Environment{},
+		Distributions: map[string]*claircore.Distribution{},
+		Repositories:  map[string]*claircore.Repository{},
+		Files:         map[string]claircore.File{},
+	}
+
+	layer, err := constructLayer(ctx, layerDigest, env.NodeScanningV4HostPath.Setting())
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if tmpErr := layer.Close(); tmpErr != nil {
+			err = tmpErr
+		}
+	}()
```

Comment:
The assignment to `err` in the deferred function - I don't think that would work unless we do something with the `err` variable. I think, we could log here, but there is no way of returning an error from that func and using it later.

Reply:
We have to check the error returned by `layer.Close`, and as that call is usually invoked as `defer layer.Close()`, the error checking is a bit more complicated. 
This is the solution I've followed [from here](https://trstringer.com/golang-deferred-function-error-handling/). 
If we do have better handling in our codebase already, please point me to it so I can implement that.

Reply:
As discussed via Slack, I've updated all instances - in tests I've ditched the defer, for the production code, I've updated to named returns in [c1b2de4](https://github.com/stackrox/stackrox/pull/12227/commits/c1b2de4dfa51d04e77199a375bddbf0bc8c78dee)

Reply:
Cool, named returns make it work, great!